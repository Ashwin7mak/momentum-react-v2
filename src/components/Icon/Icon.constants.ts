const CLASS_PREFIX = 'md-icon';

const DEFAULTS = {
  WEIGHT: 'regular',
  SCALE: 32,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
  notFound: `${CLASS_PREFIX}-not-found`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
