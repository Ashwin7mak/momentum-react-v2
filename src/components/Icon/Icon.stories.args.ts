import { ICON_CONSTANTS as CONSTANTS } from './';

export default {
  name: {
    defaultValue: 'accessibility',
    description: 'Name of the icon.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  scale: {
    defaultValue: CONSTANTS.DEFAULTS.SCALE,
    description: 'Scale represents the size/scale of te icon.',
    options: [undefined, ...Object.values(CONSTANTS.SIZES)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.SCALE,
      },
    },
  },
  autoScale: {
    defaultValue: CONSTANTS.DEFAULTS.AUTO_SCALE,
    description: 'If set to true, the icon size will match the parent size.',
    options: [undefined, true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.AUTO_SCALE,
      },
    },
  },
  weight: {
    defaultValue: CONSTANTS.DEFAULTS.WEIGHT,
    description: 'Represents the style of the icon. Note: Not all icons have all 4 styles.',
    options: [undefined, ...Object.values(CONSTANTS.WEIGHTS)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.WEIGHT,
      },
    },
  },
  className: {
    defaultValue: undefined,
    description:
      'If present, the class name will be added to the underlying svg. Used to override styles by consumers.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  color: {
    defaultValue: undefined,
    description:
      'Color for the icon to be filled and stroked with. This overrides the `fillColor` and `strokeColor` props.',
    control: { type: 'color' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
};
