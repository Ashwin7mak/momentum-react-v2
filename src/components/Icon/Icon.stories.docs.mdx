`<Icon />` is a generic icon component that acts as an dynamic icon loader.
