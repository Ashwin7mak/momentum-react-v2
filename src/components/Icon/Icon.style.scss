svg.coloured {
  fill: initial;

  path:not([fill])  {
    fill: var(--theme-text-primary-normal, black);
  }
}

.md-icon-wrapper {
  display: inline-block;

  svg {
    fill: var(--theme-text-primary-normal, black);
  }

  svg[data-scale="8"] {
    width: 0.5rem;
    height: 0.5rem;
  }

  svg[data-scale="10"] {
    width: 0.625rem;
    height: 0.625rem;
  }

  svg[data-scale="12"] {
    width: 0.75rem;
    height: 0.75rem;
  }

  svg[data-scale="14"] {
    width: 0.875rem;
    height: 0.875rem;
  }

  svg[data-scale="16"] {
    width: 1rem;
    height: 1rem;
  }

  svg[data-scale="18"] {
    width: 1.125rem;
    height: 1.125rem;
  }

  svg[data-scale="20"] {
    width: 1.25rem;
    height: 1.25rem;
  }

  svg[data-scale="22"] {
    width: 1.375rem;
    height: 1.375rem;
  }

  svg[data-scale="24"] {
    width: 1.5rem;
    height: 1.5rem;
  }

  svg[data-scale="28"] {
    width: 1.75rem;
    height: 1.75rem;
  }

  svg[data-scale="32"] {
    width: 2rem;
    height: 2rem;
  }

  svg[data-scale="36"] {
    width: 2.25rem;
    height: 2.25rem;
  }

  svg[data-scale="40"] {
    width: 2.5rem;
    height: 2.5rem;
  }

  svg[data-scale="48"] {
    width: 3rem;
    height: 3rem;
  }

  svg[data-scale="56"] {
    width: 3.5rem;
    height: 3.5rem;
  }

  svg[data-scale="64"] {
    width: 4rem;
    height: 4rem;
  }

  svg[data-scale="120"] {
    width: 7.5rem;
    height: 7.5rem;
  }

  svg[data-scale="124"] {
    width: 7.75rem;
    height: 7.75rem;
  }
}

.md-icon-not-found {
  color: var(--theme-text-primary-normal);
}
