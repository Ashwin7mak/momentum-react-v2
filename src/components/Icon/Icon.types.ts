export type IconScale =
  | 8
  | 10
  | 12
  | 14
  | 16
  | 18
  | 20
  | 22
  | 24
  | 28
  | 32
  | 36
  | 40
  | 48
  | 56
  | 64
  | 120
  | 124;
export interface Props {
  /**
   * className custom class to be able to override  the icon's CSS
   */
  className?: string;
  /**
   * Name of the icon. Should be lowercase. See Icon docs for
   * available icons.
   */
  name: string;
  /**
   * Scale represents the size/scale of te icon.
   * @default 32
   */
  scale?: IconScale;
  /**
   * If set to true, then the icon size will scale according
   * to the parent element.
   */
  autoScale?: boolean;
  /**
   * Represents the style of the icon.
   * Note: Not all icons have all 4 styles.
   */
  weight?: 'light' | 'regular' | 'bold' | 'filled';
  /**
   * Color to fill the icon. Theming is provided by default (dark/light)
   */
  fillColor?: string;
}
