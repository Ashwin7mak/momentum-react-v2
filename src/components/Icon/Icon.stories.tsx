import React, { FC } from 'react';
import { Story } from '@storybook/react';

import Icon, { IconProps } from './';
import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs';

import ThemeProvider from '../ThemeProvider';
import Documentation from './Icon.documentation.mdx';

const DocsPage: FC = () => (
  <>
    <Title />
    <Subtitle />
    <Description />
    <Documentation />
    <Primary />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  title: 'Momentum UI/Icon',
  component: Icon,
  parameters: {
    expanded: true,
    docs: {
      page: DocsPage,
    },
  },
  argTypes: {
    scale: {
      description: 'Scale represents the size/scale of te icon.',
      options: [
        8, 10, 12, 14, 16, 18, 20, 22, 24, 28, 32, 36, 40, 48, 56, 64, 120, 124,
      ],
      control: { type: 'select' },
    },
    weight: {
      description:
        'Represents the style of the icon. Note: Not all icons have all 4 styles.',
      options: ['light', 'regular', 'bold', 'filled'],
      control: { type: 'select' },
    },
  },
};

const Template: Story<IconProps> = (args) => (
  <>
    <Icon {...args} />
    <Icon {...args} weight="light" />
    <Icon {...args} weight="regular" />
    <Icon {...args} weight="bold" />
  </>
);

const Example = Template.bind({});

Example.args = {
  name: 'accessibility',
};

export { Example };
