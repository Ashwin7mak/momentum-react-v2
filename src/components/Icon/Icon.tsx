import React from 'react';
import './Icon.style.scss';
import { Props } from './Icon.types';
import { useDynamicSVGImport } from '../../hooks/useDynamicSVGImport';
import { DEFAULTS, GLYPH_NOT_FOUND, STYLE } from './Icon.constants';
import classnames from 'classnames';

/**
 * Icon component that can dynamically display SVG icons with a valid name.
 */
const Icon: React.FC<Props> = (props: Props) => {
  const { autoScale, className, color, fillColor, id, name, scale, strokeColor, style, weight } =
    props;
  const { SvgIcon, error } = useDynamicSVGImport(`${name}-${weight || DEFAULTS.WEIGHT}`);

  const isColoredIcon = name.indexOf('coloured') > 0;

  if (error) {
    return (
      <div className={STYLE.wrapper} id={id} style={style}>
        <div
          className={STYLE.notFound}
          data-scale={!autoScale && (scale || DEFAULTS.SCALE)}
          data-autoscale={autoScale || DEFAULTS.AUTO_SCALE}
        >
          {GLYPH_NOT_FOUND}
        </div>
      </div>
    );
  }

  const getColors = () => {
    if (!isColoredIcon) {
      const style: Record<string, string> = {};

      if (fillColor) {
        style.fill = fillColor;
      }

      if (strokeColor) {
        style.stroke = strokeColor;
      }

      if (color) {
        style.fill = color;
        style.stroke = color;
      }

      return style;
    }

    return null;
  };

  if (SvgIcon) {
    return (
      <div className={classnames(STYLE.wrapper, className)} id={id} style={style}>
        <SvgIcon
          // coloured class is added to avoid theming the fixed colours inside coloured icons
          className={classnames({ [STYLE.coloured]: isColoredIcon })}
          style={{ ...getColors() }}
          stroke="currentColor"
          fill="currentColor"
          viewBox="0, 0, 32, 32"
          width="100%"
          height="100%"
          data-scale={!autoScale && (scale || DEFAULTS.SCALE)}
          data-autoscale={autoScale || DEFAULTS.AUTO_SCALE}
        />
      </div>
    );
  }

  return null;
};

export default Icon;
