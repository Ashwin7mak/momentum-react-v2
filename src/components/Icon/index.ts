import { default as Icon } from './Icon';
import * as CONSTANTS from './Icon.constants';
import { Props } from './Icon.types';

export { CONSTANTS as ICON_CONSTANTS };

export type IconProps = Props;

export default Icon;
