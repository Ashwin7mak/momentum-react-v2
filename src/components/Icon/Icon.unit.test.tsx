import Icon from '.';
import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import { STYLE } from './Icon.constants';
import { act } from 'react-dom/test-utils';

const mountAndWait = async (component) => {
  const _container = mount(component);
  await waitForComponentToPaint(_container);
  return _container;
};

const waitForComponentToPaint = async (wrapper: ReactWrapper) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();
  });
};

describe('Icon', () => {
  let container;

  describe('snapshot', () => {
    it('should match snapshot', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with scale', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" scale={16} />);

      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with autoScale set to 'true'", async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" autoScale={true} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with autoScale set to a percentage value', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" autoScale={150} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with weight', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" weight="light" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with color', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" color="red" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with fillColor', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" fillColor="red" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with strokeColor', async () => {
      expect.assertions(1);

      container = await mountAndWait(<Icon name="accessibility" strokeColor="red" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with invalid name', async () => {
      expect.assertions(1);
      jest.setMock('@momentum-ui/icons-rebrand/svg/invalid_icon_name.svg', undefined);

      container = await mountAndWait(<Icon name="invalid_icon_name" />);

      expect(container).toMatchSnapshot();
      jest.dontMock('@momentum-ui/icons-rebrand/svg/invalid_icon_name.svg');
    });
  });

  describe('attributes', () => {
    it('should have its main class', async () => {
      const wrapper = await mountAndWait(<Icon name="accessibility" />);
      const icon = wrapper.find(Icon).getDOMNode();

      expect(icon.classList.contains(STYLE.wrapper));
    });

    it('should have custom class if provided', async () => {
      const testClass = 'testClass';

      const wrapper = await mountAndWait(<Icon name="accessibility" className={testClass} />);
      const svg = wrapper.find('svg').getDOMNode();

      expect(svg.classList.contains(testClass)).toBe(true);
    });

    it('should pass scale prop', async () => {
      const scale = 16;

      const wrapper = await mountAndWait(<Icon name={'accessibility'} scale={scale} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('data-scale')).toBe(`${scale}`);
    });

    it('should pass autoScale prop and disable scale prop', async () => {
      const autoScale = true;

      const wrapper = await mountAndWait(<Icon name={'accessibility'} autoScale={autoScale} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('data-autoscale')).toBe(`${autoScale}`);
      expect(icon.getAttribute('data-scale')).toBe('false');
    });

    it('should pass autoScale prop as numeric value when set appropriately', async () => {
      const autoScale = 150;

      const wrapper = await mountAndWait(<Icon name={'accessibility'} autoScale={autoScale} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('data-autoscale')).toBe(`${autoScale}`);
      expect(icon.getAttribute('data-scale')).toBe('false');
    });

    it('should pass md-icon-coloured class if name of icon contains coloured', async () => {
      const wrapper = await mountAndWait(<Icon name={'accessibility-coloured'} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.classList.contains('md-icon-coloured')).toBe(true);
    });

    it('should pass color prop', async () => {
      const color = 'blue';

      const wrapper = await mountAndWait(<Icon name={'accessibility'} color={color} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('style')).toBe(`fill: ${color}; stroke: ${color};`);
    });

    it('should pass fillColor prop', async () => {
      const fillColor = 'blue';

      const wrapper = await mountAndWait(<Icon name={'accessibility'} fillColor={fillColor} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('style')).toBe(`fill: ${fillColor};`);
    });

    it('should pass strokeColor prop', async () => {
      const strokeColor = 'blue';

      const wrapper = await mountAndWait(<Icon name={'accessibility'} strokeColor={strokeColor} />);
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('style')).toBe(`stroke: ${strokeColor};`);
    });

    it('should override fillColor and strokeColor when color is provided', async () => {
      const color = 'blue';
      const fillColor = 'green';
      const strokeColor = 'red';

      const wrapper = await mountAndWait(
        <Icon
          name={'accessibility'}
          color={color}
          fillColor={fillColor}
          strokeColor={strokeColor}
        />
      );

      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('style')).toBe(`fill: ${color}; stroke: ${color};`);
    });

    it('should not pass fillColor prop if icon is already coloured', async () => {
      const fillColor = 'blue';

      const wrapper = await mountAndWait(
        <Icon name={'accessibility-coloured'} fillColor={fillColor} />
      );
      const icon = wrapper.find('svg').getDOMNode();

      expect(icon.getAttribute('style')).toBe(null);
    });
  });
});
