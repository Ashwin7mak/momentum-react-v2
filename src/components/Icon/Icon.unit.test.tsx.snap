// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon snapshot should match snapshot 1`] = `
<Icon
  name="accessibility"
>
  <div
    className="md-icon-wrapper"
  >
    <svg
      className=""
      data-autoscale={false}
      data-scale={32}
      height="100%"
      style={null}
      viewBox="0, 0, 32, 32"
      width="100%"
    />
  </div>
</Icon>
`;

exports[`Icon snapshot should match snapshot with autoScale 1`] = `
<Icon
  autoScale={true}
  name="accessibility"
>
  <div
    className="md-icon-wrapper"
  >
    <svg
      className=""
      data-autoscale={true}
      data-scale={false}
      height="100%"
      style={null}
      viewBox="0, 0, 32, 32"
      width="100%"
    />
  </div>
</Icon>
`;

exports[`Icon snapshot should match snapshot with fillColor 1`] = `
<Icon
  fillColor="red"
  name="accessibility"
>
  <div
    className="md-icon-wrapper"
  >
    <svg
      className=""
      data-autoscale={false}
      data-scale={32}
      height="100%"
      style={
        Object {
          "fill": "red",
        }
      }
      viewBox="0, 0, 32, 32"
      width="100%"
    />
  </div>
</Icon>
`;

exports[`Icon snapshot should match snapshot with invalid name 1`] = `
<Icon
  name="invalid_icon_name"
>
  <p
    className="md-icon-not-found"
  >
    Icon not found.
  </p>
</Icon>
`;

exports[`Icon snapshot should match snapshot with scale 1`] = `
<Icon
  name="accessibility"
  scale={16}
>
  <div
    className="md-icon-wrapper"
  >
    <svg
      className=""
      data-autoscale={false}
      data-scale={16}
      height="100%"
      style={null}
      viewBox="0, 0, 32, 32"
      width="100%"
    />
  </div>
</Icon>
`;

exports[`Icon snapshot should match snapshot with weight 1`] = `
<Icon
  name="accessibility"
  weight="light"
>
  <div
    className="md-icon-wrapper"
  >
    <svg
      className=""
      data-autoscale={false}
      data-scale={32}
      height="100%"
      style={null}
      viewBox="0, 0, 32, 32"
      width="100%"
    />
  </div>
</Icon>
`;
